metaflow/events.py

```diff
@@ -0,0 +1,136 @@
+    timestamp : datetime
+        timestamp recording creation time for the event.
+    type : str
+        type for the event - one of `event` or `run`
+    """
+
+
+class Trigger(object):
+    """
+    Defines a container of event triggers' metadata.
+
+    """
+
+    def __init__(self, _meta=None):
+        if _meta is None:
+            _meta = []
+
+        _meta.sort(key=lambda x: x.get("timestamp") or float("-inf"), reverse=True)
+
+        self._runs = None
+        self._events = [
+            MetaflowEvent(
+                **{
+                    **obj,
+                    # Add timestamp as datetime. Guaranteed to exist for Metaflow
+                    # events - best effort for everything else.
+                    **(
+                        {"timestamp": datetime.fromtimestamp(obj["timestamp"])}
+                        if obj.get("timestamp")
+                        and isinstance(obj.get("timestamp"), int)
+                        else {}
+                    ),
+                }
+            )
+            for obj in _meta
+        ]
+
+    @classmethod
+    def from_runs(cls, run_objs):
+        run_objs.sort(key=lambda x: x.finished_at, reverse=True)
+        trigger = Trigger(
+            [
+                {
+                    "type": "run",
+                    "timestamp": run_obj.finished_at,
+                    "name": "metaflow.%s.%s" % (run_obj.parent.id, run_obj["end"].id),
+                    "id": run_obj.end_task.pathspec,
+                }
+                for run_obj in run_objs
+            ]
+        )
+        trigger._runs = run_objs
+        return trigger
+
+    @property
+    def event(self):
+        """
+        Returns a `MetaflowEvent` object corresponding to the triggering event. If multiple events triggered the run, returns the latest event.
+        """
+        return next(iter(self.events), None)
+
+    @property
+    def events(self):
+        """
+        Returns a list of `MetaflowEvent` objects correspondings to all the triggering events.
+        """
+        return list(self._events) or None
+
+    @property
+    def run(self):
+        """
+        If the triggering event is a Metaflow run, returns the corresponding `Run` object. If multiple runs triggered the run, returns the latest run. `None` if the event is not a `Run`.
+        """
+        return next(iter(self.runs), None)
+
+    @property
+    def runs(self):
+        """
+        If the triggering events correspond to Metaflow runs, returns a list of `Run` objects. Otherwise returns `None`.
+        """
+        if self._runs is None:
+            # to avoid circular import
+            from metaflow import Run
+
+            self._runs = [
+                Run(
+                    # object id is the task pathspec for events that map to run
+                    obj.id[: obj.id.index("/", obj.id.index("/") + 1)],
+                    _namespace_check=False,
+                )
+                for obj in filter(
```

Comment:
simpler: `for obj in self.events if obj.type == "run"`

Reply:
To avoid a copy (cheap but why), just do `for obj in self._events`. It also will work if empty events (this code won't since filter over None will break)